### PR TITLE
Fix bug: Process.OutputDataReceived not firing 

### PR DESCRIPTION
**Pull request by MarcoRossignoli**
fixes https://github.com/dotnet/corefx/issues/33746

/cc @wtgodbe, @krwq @stephentoub 

Aside issue, I get msbuild error when try to run on linux
```
marco@Ubuntu1404:~/repos/corefx/src/System.Diagnostics.Process$ ../../Tools/msbuild.sh  System.Diagnostics.Process.sln /t:rebuildandtest /v:m
../../Tools/msbuild.sh: line 4: /home/marco/repos/corefx/Tools/dotnetcli/dotnet: No such file or directory
```

### CODE REVIEW

**Comment by stephentoub** [comment]
Thanks. I expect this is going to lead to sporadic failures, as the read loop may still be in a ReadAsync when the stream is closed. I think we probably still need to keep the CTS and use it to cancel that read when Close/Dispose is called.

**Comment by MarcoRossignoli** [comment]
> Thanks. I expect this is going to lead to sporadic failures, as the read loop may still be in a ReadAsync when the stream is closed. I think we probably still need to keep the CTS and use it to cancel that read when Close/Dispose is called.

Understood, you mean a new method on `AsyncStreamReader.Close/Dispose` to call on Close/Dispose or an overload on `CancelOperation(bool stopReader = false)` or do you have other ideas(cleanup and set _output to null after `CancelOutputRead`, but I don't like it)?

**Comment by stephentoub** [comment]
> you mean a new method on AsyncStreamReader.Close/Dispose to call on Close/Dispose 

Yup :)

**Comment by MarcoRossignoli** [comment]
@stephentoub PTAL

**Comment by krwq** [comment]
@MarcoRossignoli 

> Aside issue, I get msbuild error when try to run on linux

use this instead:

```
alias msbuild="<corefx_root>/.dotnet/dotnet msbuild"
```

**Comment by MarcoRossignoli** [comment]
> use this instead:

@krwq it works thank's!

**Comment by ViktorHofer** [comment]
Or just corefx/eng/common/msbuild.sh

**Comment by MarcoRossignoli** [comment]
@dotnet-bot test Linux arm64 Release Build please
@dotnet-bot test Linux x64 Release Build please

**Comment by MarcoRossignoli** [comment]
@ViktorHofer folder name "eng" is acronym of?

**Comment by stephentoub** [comment]
> folder name "eng" is acronym of?

It's an abbreviation of engineering.

**Comment by MarcoRossignoli** [comment]
"Packaging All Configurations x64 Debug Build" CI is broken but I don't think is related to this PR.

**Comment by ViktorHofer** [comment]
test Packaging All Configurations x64 Debug Build

**Comment by MarcoRossignoli** [comment]
@ViktorHofer my colleague dotnet-bot is now implicit?

**Comment by ViktorHofer** [comment]
correct :)

File: src/System.Diagnostics.Process/tests/ProcessStreamReadTests.cs [hunk 1]
```diff
@@ -88,6 +89,53 @@ public void TestAsyncOutputStream()
...
+                int minDataArrived = 5;
+                p.StartInfo.RedirectStandardOutput = true;
+                p.OutputDataReceived += (s, e) =>
+                {
+                    if(!string.IsNullOrEmpty(e.Data))
+                    { 
+                        dataReceived.Add(int.Parse(e.Data));
+                    }
+
+                    if(dataReceived.Count > minDataArrived)
```
Comment (stephentoub) [inline src/System.Diagnostics.Process/tests/ProcessStreamReadTests.cs]: Nit: please use a space between the ifs and the opening parens, here and above

File: src/System.Diagnostics.Process/tests/ProcessStreamReadTests.cs [hunk 2]
```diff
@@ -88,6 +89,53 @@ public void TestAsyncOutputStream()
             }
         }
 
+        [Fact]
+        // On this test RemotelyInvokable.WriteLoop never returns, we cannot use RemotelyInvokable.WriteLoopUapCmd workaround
+        [SkipOnTargetFramework(TargetFrameworkMonikers.Uap)]
+        public void TestAsyncOutputStream_BeginCancelBegin_OutputReadLine()
```
Comment (stephentoub) [inline src/System.Diagnostics.Process/tests/ProcessStreamReadTests.cs]: This should be Outerloop, given the long sleeps involved.

File: src/System.Diagnostics.Process/tests/ProcessStreamReadTests.cs [hunk 3]
```diff
@@ -88,6 +89,53 @@ public void TestAsyncOutputStream()
...
+                p.CancelOutputRead();
+
+                // Wait to drain async stream queue
+                Thread.Sleep(3000);
+
+                minDataArrived = 10;
+                p.BeginOutputReadLine();
+
+                Assert.True(dataArrivedEvent.WaitOne(WaitInMS), "Data not come after CancelOutputRead");
+                Assert.True(dataReceived.Count >= minDataArrived);
```
Comment (stephentoub) [inline src/System.Diagnostics.Process/tests/ProcessStreamReadTests.cs]: I'm not understanding what the test is actually verifying:
- there are race conditions here that could cause this to be true even if it's not behaving as expected
- there's nothing verifying that the callback isn't invoked when it's supposed to not be
- there's nothing verifying that data that was supposed to be thrown away is thrown away
etc.

File: src/System.Diagnostics.Process/tests/ProcessStreamReadTests.cs (hunk 2 above)
Comment (MarcoRossignoli) [reply src/System.Diagnostics.Process/tests/ProcessStreamReadTests.cs]: test now run for no more than 3.x seconds...do we run in outerloop  again?

Comment (stephentoub) [reply src/System.Diagnostics.Process/tests/ProcessStreamReadTests.cs]: Yes, tests that run for seconds should be Outerloop.

File: src/System.Diagnostics.Process/tests/ProcessStreamReadTests.cs (hunk 3 above)
Comment (MarcoRossignoli) [reply src/System.Diagnostics.Process/tests/ProcessStreamReadTests.cs]: I rewrote test, the only "weak part" I think is 3 seconds sleep to be sure to lose data, I don't find a better way to be sure that queue has drained data, is async and it depends on non deterministic factors, do you have a better idea?

Comment (MarcoRossignoli) [reply src/System.Diagnostics.Process/tests/ProcessStreamReadTests.cs]: I moved callback out of `RemotelyInvokable.cs` because we don't run test on UAP.

File: src/System.Diagnostics.Process/tests/ProcessStreamReadTests.cs (hunk 2 above)
Comment (MarcoRossignoli) [reply src/System.Diagnostics.Process/tests/ProcessStreamReadTests.cs]: done

**Comment by MarcoRossignoli** [comment]
@dotnet-bot test Outerloop Windows x64 Debug Build
@dotnet-bot test Outerloop Linux x64 Debug Build
@dotnet-bot test Outerloop NETFX x86 Debug Build
@dotnet-bot test Outerloop UWP CoreCLR x64 Debug Build
@dotnet-bot test Outerloop Linux x64 Release Build please
@dotnet-bot test Outerloop Windows x86 Release Build please

**Comment by stephentoub** [comment]
@dotnet-bot help

File: src/System.Diagnostics.Process/tests/ProcessStreamReadTests.cs (hunk 3 above)
Comment (MarcoRossignoli) [reply src/System.Diagnostics.Process/tests/ProcessStreamReadTests.cs]: :sob: named EventWaitHandle is not supported everywhere.
I'll try other solution...without sync primitive it's not so easy write a strong test(with only Thread.Sleep)...do you know on codebase some similar test needs?

Comment (MarcoRossignoli) [reply src/System.Diagnostics.Process/tests/ProcessStreamReadTests.cs]: I could use "shared file+acquire open polling/sleep" to syncronize process on all plat, should be supported everywhere.

Comment (MarcoRossignoli) [reply src/System.Diagnostics.Process/tests/ProcessStreamReadTests.cs]: Previous version with some changes, what do you think?
Final assertion on gaps should ensure:
> there's nothing verifying that the callback isn't invoked when it's supposed to not be
there's nothing verifying that data that was supposed to be thrown away is thrown away

and new `Wait 'WaitInMS' seconds after last signaled data received` should ensure
> there are race conditions here that could cause this to be true even if it's not behaving as expected
```cs
        [OuterLoop]
        [Fact]
        // On this test RemotelyInvokable.WriteLoop never returns, we cannot use RemotelyInvokable.WriteLoopUapCmd workaround
        [SkipOnTargetFramework(TargetFrameworkMonikers.Uap)]
        public void TestAsyncOutputStream_BeginCancelBegin_OutputReadLine()
        {
            ConcurrentBag<int> dataReceived = new ConcurrentBag<int>();

            var dataReceivedEvent = new AutoResetEvent(false);

            Process p = CreateProcessPortable(() => 
            {
                int counter = 0;

                for (; ; )
                {
                    Thread.Sleep(500);
                    Console.WriteLine(++counter);
                }
            });

            p.StartInfo.RedirectStandardOutput = true;
            p.OutputDataReceived += (s, e) =>
            {
                if (e.Data != null)
                {
                    Console.WriteLine(e.Data);
                    dataReceived.Add(int.Parse(e.Data));
                    dataReceivedEvent.Set();
                }
            };

            try
            {
                // Start process
                p.Start();

                // Start listening
                p.BeginOutputReadLine();

                // Wait some seconds to accumulate data
                Thread.Sleep(5000);

                // Wait for data signal
                Assert.True(dataReceivedEvent.WaitOne(WaitInMS), "Data not received after BeginOutputReadLine()");

                // Suspend listening
                p.CancelOutputRead();

                // Wait 'WaitInMS' seconds after last signaled data received
                DateTime waitStart = DateTime.UtcNow;
                while (dataReceivedEvent.WaitOne(WaitInMS))
                {
                    Thread.Sleep(1000);

                    // if we'are waiting for more than 'WaitInMS' plus some seconds means that we haven't correctly stopped output
                    Assert.True(DateTime.UtcNow < waitStart.AddMilliseconds(WaitInMS + 5000));
                }

                // We assert that we've received all counter data, with no gaps
                Assert.Empty(Enumerable.Range(dataReceived.OrderBy(v => v).First(), dataReceived.OrderBy(v => v).Last()).Except(dataReceived));

                // Start listening
                p.BeginOutputReadLine();

                // Wait some seconds to accumulate new data
                Thread.Sleep(5000);

                // Wait for data signal
                Assert.True(dataReceivedEvent.WaitOne(WaitInMS), "Data not received after CancelOutputRead()/BeginOutputReadLine()");

                // We assert that we've received new data but we have a gap in serie due to lost number between CancelOutputRead() and BeginOutputReadLine()
                Assert.NotEmpty(Enumerable.Range(dataReceived.OrderBy(v => v).First(), dataReceived.OrderBy(v => v).Last()).Except(dataReceived));
            }
            finally
            {
               p.Kill();
            }
        }

```
test output
```
1
2
3
4
5
6
7
8
9
70
71
72
73
74
75
76
77
78
79
```

File: src/System.Diagnostics.Process/tests/ProcessStreamReadTests.cs [hunk 4]
```diff
@@ -88,6 +88,237 @@ public void TestAsyncOutputStream()
             }
         }
 
+        [Fact]
+        // On this test RemotelyInvokable.WriteLoop never returns, we cannot use RemotelyInvokable.WriteLoopUapCmd workaround
+        [SkipOnTargetFramework(TargetFrameworkMonikers.Uap)]
+        public void TestAsyncOutputStream_BeginCancelBegin_OutputReadLine()
+        {
+            // We use file system lock tecnique to sync process
```
Comment (danmosemsft) [inline src/System.Diagnostics.Process/tests/ProcessStreamReadTests.cs]: nit: 'technique'

File: src/System.Diagnostics.Process/tests/ProcessStreamReadTests.cs [hunk 5]
```diff
@@ -88,6 +88,237 @@ public void TestAsyncOutputStream()
...
+            // We cannot use named wait event because are not supported on all platform
+
+            Dictionary<string, FileSystemLock> parentLockList = TestAsyncOutputStream_BeginCancelBegin_OutputReadLine_LockHelper();
+
+            List<int> dataReceived = new List<int>();
+            var dataArrivedEvent = new AutoResetEvent(false);
+            var dataConfirmedEvent = new AutoResetEvent(false);
+
+            Process p = CreateProcessPortable(TestAsyncOutputStream_BeginCancelBegin_OutputReadLine_RemotelyInvokable);
```
Comment (danmosemsft) [inline src/System.Diagnostics.Process/tests/ProcessStreamReadTests.cs]: using statement?

**Comment by MarcoRossignoli** [comment]
@stephentoub PTAL, after test with `EventWaitHandle`(https://github.com/dotnet/corefx/pull/33974/commits/f30f23b33c9d881d1a7ffd75af94944174658df4) not supported on all plat I opted for custom file system lock to sync process and have a deterministic asserts. Let me know!

**Comment by MarcoRossignoli** [comment]
test Windows x86 Release Build

**Comment by karelz** [comment]
@danmosemsft @stephentoub looks like your feedback was addressed (based on brief scan of outdated comments). Are you ok with merging?

File: src/System.Diagnostics.Process/tests/ProcessStreamReadTests.cs [hunk 6]
```diff
@@ -88,6 +88,243 @@ public void TestAsyncOutputStream()
...
+        // On this test RemotelyInvokable.WriteLoop never returns, we cannot use RemotelyInvokable.WriteLoopUapCmd workaround
+        [SkipOnTargetFramework(TargetFrameworkMonikers.Uap)]
+        public void TestAsyncOutputStream_BeginCancelBegin_OutputReadLine()
+        {
+            // We use file system lock technique to sync process
+            // We cannot use named wait event because are not supported on all platform
+
+            Dictionary<string, FileSystemLock> parentLockList = TestAsyncOutputStream_BeginCancelBegin_OutputReadLine_LockHelper(nameof(TestAsyncOutputStream_BeginCancelBegin_OutputReadLine));
+
+            List<int> dataReceived = new List<int>();
```
Comment (danmosemsft) [inline src/System.Diagnostics.Process/tests/ProcessStreamReadTests.cs]: nit, this could be var.

File: src/System.Diagnostics.Process/tests/ProcessStreamReadTests.cs [hunk 7]
```diff
@@ -88,6 +88,243 @@ public void TestAsyncOutputStream()
...
+            var dataArrivedEvent = new AutoResetEvent(false);
+            var dataConfirmedEvent = new AutoResetEvent(false);
+
+            using(Process p = CreateProcessPortable(TestAsyncOutputStream_BeginCancelBegin_OutputReadLine_RemotelyInvokable))
```
Comment (danmosemsft) [inline src/System.Diagnostics.Process/tests/ProcessStreamReadTests.cs]: nit, space after `using`

File: src/System.Diagnostics.Process/tests/ProcessStreamReadTests.cs [hunk 8]
```diff
@@ -88,6 +88,243 @@ public void TestAsyncOutputStream()
...
+        private class FileSystemLock : IDisposable
+        {
+            private readonly string _lockFileName = "";
+            private FileStream _fs = null;
+
+            public FileSystemLock(string lockName, string directoryName)
+            {
+                string rootDirectory = Path.Combine(Path.GetTempPath(), directoryName);
+
+                if (!Directory.Exists(rootDirectory))
```
Comment (danmosemsft) [inline src/System.Diagnostics.Process/tests/ProcessStreamReadTests.cs]: No need to change this, but note that there is no need to test for existence, as CreateDirectory will just return.

File: src/System.Diagnostics.Process/tests/ProcessStreamReadTests.cs [hunk 9]
```diff
@@ -88,6 +88,243 @@ public void TestAsyncOutputStream()
...
+                {
+                    Directory.CreateDirectory(rootDirectory);
+                }
+
+                _lockFileName = Path.Combine(rootDirectory, lockName);
+            }
+
+            public bool WaitSignal(int millisecondsTimeout)
+            {
+                DateTime maxWait = DateTime.UtcNow.AddMilliseconds(millisecondsTimeout);
```
Comment (danmosemsft) [inline src/System.Diagnostics.Process/tests/ProcessStreamReadTests.cs]: Nit, no need to change: when timing millisecond level, it would be more idiomatic (and simpler) to use Stopwatch.

File: src/System.Diagnostics.Process/tests/ProcessStreamReadTests.cs [hunk 10]
```diff
@@ -88,6 +88,243 @@ public void TestAsyncOutputStream()
...
+                catch (IOException)
+                {
+                    return true;
+                }
+            }
+
+            public bool TryAcquire(int millisecondsTimeout)
+            {
+                DateTime maxWait = DateTime.UtcNow.AddMilliseconds(millisecondsTimeout);
+                for (; ; )
```
Comment (danmosemsft) [inline src/System.Diagnostics.Process/tests/ProcessStreamReadTests.cs]: `while (true)` is more idiomatic c#.

File: src/System.Diagnostics.Process/src/System/Diagnostics/AsyncStreamReader.cs [hunk 11]
```diff
@@ -67,6 +68,11 @@ internal AsyncStreamReader(Stream stream, Action<string> callback, Encoding enco
         // User calls BeginRead to start the asynchronous read
         internal void BeginReadLine()
         {
+            if (_cancelOperation)
```
[the remark this answers is not in the document]
Comment (MarcoRossignoli) [reply src/System.Diagnostics.Process/src/System/Diagnostics/AsyncStreamReader.cs]: simply to be consistent with full framework https://referencesource.microsoft.com/#System/services/monitoring/system/diagnosticts/AsyncStreamReader.cs,135 it's not `volatile` so we could just set it.

File: src/System.Diagnostics.Process/tests/ProcessStreamReadTests.cs (hunk 8 above)
Comment (MarcoRossignoli) [reply src/System.Diagnostics.Process/tests/ProcessStreamReadTests.cs]: you're right, less code is better I'll fix it

File: src/System.Diagnostics.Process/src/System/Diagnostics/AsyncStreamReader.cs (hunk 11 above)
[the remark this answers is not in the document]
Comment (MarcoRossignoli) [reply src/System.Diagnostics.Process/src/System/Diagnostics/AsyncStreamReader.cs]: fixed thank's!

File: src/System.Diagnostics.Process/tests/ProcessStreamReadTests.cs (hunk 7 above)
Comment (MarcoRossignoli) [reply src/System.Diagnostics.Process/tests/ProcessStreamReadTests.cs]: fixed thank's!

File: src/System.Diagnostics.Process/tests/ProcessStreamReadTests.cs (hunk 6 above)
Comment (MarcoRossignoli) [reply src/System.Diagnostics.Process/tests/ProcessStreamReadTests.cs]: fixed thank's!

File: src/System.Diagnostics.Process/tests/ProcessStreamReadTests.cs (hunk 9 above)
Comment (MarcoRossignoli) [reply src/System.Diagnostics.Process/tests/ProcessStreamReadTests.cs]: fixed thank's!

File: src/System.Diagnostics.Process/tests/ProcessStreamReadTests.cs (hunk 10 above)
Comment (danmosemsft) [reply src/System.Diagnostics.Process/tests/ProcessStreamReadTests.cs]: Both make the same code (or should - they do in latest Roslyn/RyuJIT). It is a matter of style. I only want to be consistent. In CoreFX I see 279 of the "for" format and 485 of the "while" format so I would use the latter.

Comment (MarcoRossignoli) [reply src/System.Diagnostics.Process/tests/ProcessStreamReadTests.cs]: Sounds good Dan!I'll fix it ASAP. Thank's for your time!

Comment (MarcoRossignoli) [reply src/System.Diagnostics.Process/tests/ProcessStreamReadTests.cs]: Fixed

File: src/System.Diagnostics.Process/tests/ProcessStreamReadTests.cs [hunk 12]
```diff
@@ -88,6 +88,240 @@ public void TestAsyncOutputStream()
...
+                }
+                catch (IOException)
+                {
+                    return true;
+                }
+            }
+
+            public bool TryAcquire(int millisecondsTimeout)
+            {
+                DateTime maxWait = DateTime.UtcNow.AddMilliseconds(millisecondsTimeout);
```
Comment (MarcoRossignoli) [inline src/System.Diagnostics.Process/tests/ProcessStreamReadTests.cs]: My track: I have to amend here(on mobile now)

**Comment by MarcoRossignoli** [comment]
test Linux arm64 Release Build please

**Comment by MarcoRossignoli** [comment]
test Linux arm64 Release Build please

**Comment by MarcoRossignoli** [comment]
test Windows x86 Release Build  please
test Linux x64 Release Build please

**Comment by danmosemsft** [comment]
@stephentoub do you have any remaining feedback?

**Comment by stephentoub** [comment]
I got your pings. Just haven't yet reviewed.

File: src/System.Diagnostics.Process/tests/ProcessStreamReadTests.cs [hunk 13]
```diff
@@ -88,6 +88,240 @@ public void TestAsyncOutputStream()
...
+                { "childProduced456", new FileSystemLock("childProduced456", directoryName) },
+                { "childProduced789", new FileSystemLock("childProduced789", directoryName) },
+                { "shutdownChildProcess", new FileSystemLock("shutdownChildProcess", directoryName) }
+            };
+        }
+
+        /// <summary>
+        /// We save lock files in temp directory
+        /// </summary>
+        private class FileSystemLock : IDisposable
```
Comment (stephentoub) [inline src/System.Diagnostics.Process/tests/ProcessStreamReadTests.cs]: Could anonymous or named pipes be used instead of a custom primitive like this? A signaler can write a but to the pipe and a waiter can read a byte.

Comment (MarcoRossignoli) [reply src/System.Diagnostics.Process/tests/ProcessStreamReadTests.cs]: ok I'll try with pipes!Thank's!

Comment (MarcoRossignoli) [reply src/System.Diagnostics.Process/tests/ProcessStreamReadTests.cs]: @stephentoub PTAL I used pipes.

File: src/System.Diagnostics.Process/tests/ProcessStreamReadTests.cs [hunk 14]
```diff
@@ -88,6 +90,120 @@ public void TestAsyncOutputStream()
             }
         }
 
+        [Fact]
```
Comment (stephentoub) [inline src/System.Diagnostics.Process/tests/ProcessStreamReadTests.cs]: As we're explicitly sleeping for non-trivial periods of time, please mark this as [Outerloop]

File: src/System.Diagnostics.Process/tests/ProcessStreamReadTests.cs [hunk 15]
```diff
@@ -88,6 +90,120 @@ public void TestAsyncOutputStream()
...
+                    Assert.True(WaitPipeSignal(pipeRead, WaitInMS), "Missing child signal for value 1");
+                    Assert.True(dataArrivedEvent.WaitOne(WaitInMS), "Value 1 not received");
+
+                    //Stop listening and signal to produce value 2
+                    p.CancelOutputRead();
+                    pipeWrite.WriteByte(0);
+
+                    // Wait child signal produce number 2
+                    Assert.True(WaitPipeSignal(pipeRead, WaitInMS), "Missing child signal for value 2");
+                    // We need to sleep to be sure to drain async queue
```
Comment (stephentoub) [inline src/System.Diagnostics.Process/tests/ProcessStreamReadTests.cs]: What happens if things are delayed and we don't actually completely drain the queue in this 500ms?  Will the test spuriously fail, or will the test simply not end up validating what it was intended to validate?  I'm ok with the latter, but we'll need to fix the former if that's the case.

Comment (MarcoRossignoli) [reply src/System.Diagnostics.Process/tests/ProcessStreamReadTests.cs]: > Will the test spuriously fail, or will the test simply not end up validating what it was intended to validate?

Yes I'm aware of this and I expected your concern, but isn't too risky "not end up validating what it was intended to validate"? With future changes this test could became unuseful(fail forever), right?
I thought to wait for more time, classic 30 seconds or put test in a loop with some different wait time(to speed up test) , for instance wait 1,3,6,10 seconds and if fails every times fail test.
What do you think?

**Comment by MarcoRossignoli** [comment]
> Sorry for the delay. 

No problem

File: src/System.Diagnostics.Process/tests/ProcessStreamReadTests.cs (hunk 15 above)
Comment (stephentoub) [reply src/System.Diagnostics.Process/tests/ProcessStreamReadTests.cs]: > What do you think?

I expect we'd still see failures in CI, and if it fails, it won't be clear whether that's because there was a bug in the library or the test was just flaky, making the test irrelevant.

I'm not clear why the non-determinism is needed, though.  Seems like you could have two tests:
- One to validate that cancellation causes the resulting data to be thrown away
- One to validate that additional data arrives after reading is restarted

The former would synchronize between the parent and child by having the child send data and the parent read it all, then canceling and then signaling to the child to send more, then the child exiting, and the parent validating that it saw the stream end without having received additional data from the child.

The latter would do something similar.  The parent would read the data from the child, then call cancel, then signal for the child to send more, then start reading again, then signal to the child to send more, and then validate that it did in fact receive more... and we don't need to tell the difference between the data in the interim between the cancellation and the restart, and after the restart, because the first test validated we don't get data from the former.

File: src/System.Diagnostics.Process/tests/ProcessStreamReadTests.cs [hunk 16]
```diff
@@ -88,6 +90,197 @@ public void TestAsyncOutputStream()
...
+                    p.CancelOutputRead();
+                    await pipeWrite.WriteAsync(new byte[1], 0, 1);
+
+                    // Wait child signal produce number 2
+                    Assert.True(await WaitPipeSignal(pipeRead, WaitInMS), "Missing child signal for value 2");
+
+                    // Wait child process close
+                    Assert.True(p.WaitForExit(WaitInMS), "Child process didn't close");
+
+                    Assert.Equal(1, dataReceived.Count);
```
Comment (MarcoRossignoli) [inline src/System.Diagnostics.Process/tests/ProcessStreamReadTests.cs]: @stephentoub first test, no data coming after cancel.

File: src/System.Diagnostics.Process/tests/ProcessStreamReadTests.cs [hunk 17]
```diff
@@ -88,6 +90,197 @@ public void TestAsyncOutputStream()
...
+                    dataReceived.Clear();
+
+                    // Produce new value
+                    await pipeWrite.WriteAsync(new byte[1], 0, 1);
+
+                    // Wait child signal produce number 2
+                    Assert.True(await WaitPipeSignal(pipeRead, WaitInMS), "Missing child signal for value 2");
+
+                    // Stop listen
+                    p.BeginOutputReadLine();
```
Comment (MarcoRossignoli) [inline src/System.Diagnostics.Process/tests/ProcessStreamReadTests.cs]: @stephentoub my doubt is here(the reason for wait expiring of dataArrivedEvent), if child produce value and signal but there is a delay it's possible that after begin we receive 2 values and not only 1 because we receive values in async way. I would add ` Assert.False(dataArrivedEvent.WaitOne(WaitInMS), "We don't expect data come in");` before `BeginOutputReadLine` to be sure, I think we cannot have determinism because we don't know how much we need to "elaborate"(skip in case of Cancel state) data coming from child, the best we can is wait for some time(dataArrivedEvent event in this case), tell me your idea.

Comment (stephentoub) [reply src/System.Diagnostics.Process/tests/ProcessStreamReadTests.cs]: Why does it matter how much we need to skip? The other test validates the correct behavior around not getting data while canceled, right? So this one then just needs to validate that we receive the data after starting up again; it doesn't matter if we also receive data during the perceived cancellation window.

Comment (MarcoRossignoli) [reply src/System.Diagnostics.Process/tests/ProcessStreamReadTests.cs]: > Why does it matter how much we need to skip? The other test validates the correct behavior around not getting data while canceled, right? 

Yes

> So this one then just needs to validate that we receive the data after starting up again; it doesn't matter if we also receive data during the perceived cancellation window.

OK, I let me rephrase your idea, this second test should only verify that if I call `p.CancelOutputRead();` and after `p.BeginOutputReadLine();` after second one I MUST receive some produced data(wait child exit), because first test verify that after `p.CancelOutputRead();` we don't receive data anymore, produce data between cancel/begin doens't make a lot of sense and doesn't have deterministic behaviour, thank's for your perseverance :relaxed:.

Comment (stephentoub) [reply src/System.Diagnostics.Process/tests/ProcessStreamReadTests.cs]: > after second one I MUST receive some produced data

Yes.  More specifically, you must receive all of the data that you know the child produced after being told to send more:
- Parent starts reading, then signals for the child to data.
- Child sends elements 1, 2, and 3, then waits.
- Parent reads elements 1, 2, 3, verifies it got them, then cancels, then signals to the child that it canceled.
- Child receives cancellation signal and sends next three elements, 4, 5, and 6, then waits.
- Parent immediately (or with a very short delay, doesn't really matter) re-starts reading, and signals to child.
- Child receives signal and sends next three elements, 7, 8, and 9, the exits.
- Parent reads remainder of input, and validates that it at least received 7, 8, and 9 (it might also receive 6, or 5 and 6, or 4 and 5 and 6... that's fine; the other test validated that it doesn't receive input while canceled).

Comment (MarcoRossignoli) [reply src/System.Diagnostics.Process/tests/ProcessStreamReadTests.cs]: > (it might also receive 6, or 5 and 6, or 4 and 5 and 6... that's fine; the other test validated that it doesn't receive input while canceled).

My concern was here...I rework a test with your precise steps, thank's a lot!

Comment (MarcoRossignoli) [reply src/System.Diagnostics.Process/tests/ProcessStreamReadTests.cs]: @stephentoub  did some change, now I'm using BlockingCollection Set/Wait are flaky(very fast set I'll lose some signal).

**Comment by MarcoRossignoli** [comment]
test this please

File: src/System.Diagnostics.Process/tests/ProcessStreamReadTests.cs [hunk 18]
```diff
@@ -88,6 +91,224 @@ public void TestAsyncOutputStream()
...
+
+                    // Stop listening and signal to produce value 2
+                    p.CancelOutputRead();
+                    await pipeWrite.WriteAsync(new byte[1], 0, 1);
+
+                    // Wait child signal produce number 2
+                    Assert.True(await WaitPipeSignal(pipeRead, WaitInMS), "Missing child signal for value 2");
+
+                    // Wait child process close to be sure that output buffer has been flushed
+                    Assert.True(p.WaitForExit(WaitInMS), "Child process didn't close");
```
Comment (MarcoRossignoli) [inline src/System.Diagnostics.Process/tests/ProcessStreamReadTests.cs]: > The former would synchronize between the parent and child by having the child send data and the parent read it all, then canceling and then signaling to the child to send more, then the child exiting, and the parent validating that it saw the stream end without having received additional data from the child.

@stephentoub I'm not sure about this, what if child exits but signal 2 already arrived on parent but there a delay in "parent side processing", we could receive data also after asserts, right?

Comment (MarcoRossignoli) [reply src/System.Diagnostics.Process/tests/ProcessStreamReadTests.cs]: ping @stephentoub

**Comment by MarcoRossignoli** [comment]
@stephentoub I left last doubt https://github.com/dotnet/corefx/pull/33974#discussion_r252756951 if for you it's ok maybe we could merge

File: src/System.Diagnostics.Process/tests/ProcessStreamReadTests.cs (hunk 18 above)
Comment (stephentoub) [reply src/System.Diagnostics.Process/tests/ProcessStreamReadTests.cs]: Hmm.  I guess the key issue is that there's no API exposed that let's the test "see the stream end", as I suggested.  I'll need to think about it.  In the meantime, I don't want to hold up the product fix.  Why don't you just add a comment that there's a race condition such that the test might have some false negatives.

Comment (MarcoRossignoli) [reply src/System.Diagnostics.Process/tests/ProcessStreamReadTests.cs]: I think false positives, because we cancel output and ask to child to send send new one, after that we wait for process exit and count dataReceived(expected unchanged), but if we've a bug in implementation the message could be "in processing" on parent side but we passed tests. BTW added comments.

Comment (stephentoub) [reply src/System.Diagnostics.Process/tests/ProcessStreamReadTests.cs]: I think we're saying the same thing but using opposite polarity. Just to be sure, when I say it may have false negatives, I mean the test may fail to catch a bug (to me a false positive is the test saying there's a bug when there isn't one).

Comment (MarcoRossignoli) [reply src/System.Diagnostics.Process/tests/ProcessStreamReadTests.cs]: ah ok :smile: I'll change the polarity on comment
